Add   HTTPProxyCheck to preflight checks for 'kubeadm join' subcommand

bart0sh opened the pull request:
**What this PR does / why we need it:**

Add HTTPProxyCheck for API servers
    
It makes sense to check API servers and print warnings if they're
going to be accessed through proxy. This is similar to what's
already done for 'kubeadm init'.

kad commented:
Please split independent functionality changes and bugfix into separate PRs.

/ok-to-test

kad commented:
/assign kad

bart0sh commented:
/release-note-none

bart0sh committed:
Add HTTPProxyCheck for API servers

It makes sense to check all API servers mentioned in the command
line and print warnings if they're going to be accessed through proxy.
This is similar to what's already done for 'kubeadm init'.

prateekgogia commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -1002,6 +1003,17 @@ func RunJoinNodeChecks(execer utilsexec.Interface, cfg *kubeadmapi.NodeConfigura
+	for _, server := range cfg.DiscoveryTokenAPIServers {
+		ipstr, portstr, err := net.SplitHostPort(server)
+		if err == nil {
+			if port, err := strconv.Atoi(portstr); err == nil {
```
I am trying to understand, why can't this `if` check be part of the existing loop below (originally on line 1005)?

bart0sh replied:
I did it for better readability. Existing loop adds FileContentCheck only  once. It's possible to add this code there, but we'll have to skip adding FileContentCheck if it's already added. This would make the code in the loop less readable in my opinion.

bart0sh replied:
I could only come up with this, which I consider less readable. Please, let me know if you prefer this or have better suggestion.

```
        fccheckAdded := false
        for _, server := range cfg.DiscoveryTokenAPIServers {
                ipstr, portstr, err := net.SplitHostPort(server)
                if err == nil {
                        if port, err := strconv.Atoi(portstr); err == nil {
                                checks = append(checks,
                                        HTTPProxyCheck{Proto: "https", Host: ipstr, Port: port},
                                )
                        }
                        if !fccheckAdded {
                                if ip := net.ParseIP(ipstr); ip != nil {
                                        if ip.To4() == nil && ip.To16() != nil {
                                                checks = append(checks,
                                                        FileContentCheck{Path: bridgenf6, Content: []byte{'1'}},
                                                )
                                                fccheckAdded = true // Ensure that check is added only once
                                        }
                                }
                        }
                }
        }
```

bart0sh commented:
/area kubeadm

kad commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -999,19 +999,27 @@ func RunJoinNodeChecks(execer utilsexec.Interface, cfg *kubeadmapi.NodeConfigura
+			checks = append(checks,
+				HTTPProxyCheck{Proto: "https", Host: ipstr},
+			)
+			if bridgenf6Check != nil {
```
Isn't it here it should be `if bridgenf6Check == nil` ?

bart0sh replied:
My bad. Should be fixed now.

bart0sh commented:
@kad Would you mind lgtm-ing this :) ?

kad commented:
/lgtm

timothysc reviewed (approved):
/approve

bart0sh commented:
/retest pull-kubernetes-bazel-test

bart0sh commented:
/test pull-kubernetes-bazel-test
